File: src/main/java/net/dontdrinkandroot/gitki/wicket/page/BrowsePage.java
package net.dontdrinkandroot.gitki.wicket.page;

import net.dontdrinkandroot.gitki.model.AbstractPath;
import net.dontdrinkandroot.gitki.service.git.GitService;
import net.dontdrinkandroot.gitki.wicket.GitkiWebSession;
import net.dontdrinkandroot.gitki.wicket.component.PathBreadcrumb;
import net.dontdrinkandroot.gitki.wicket.model.AbstractPathAbsoluteStringModel;
import org.apache.wicket.markup.repeater.RepeatingView;
import org.apache.wicket.model.IModel;
import org.apache.wicket.request.mapper.parameter.PageParameters;
import org.apache.wicket.spring.injection.annot.SpringBean;

/**
 * @author Philip Washington Sorst <philip@sorst.net>
 */
public class BrowsePage<T extends AbstractPath> extends DecoratorPage<T>
{
    @SpringBean
    private GitService gitService;

    public BrowsePage(PageParameters parameters)
    {
        super(parameters);
        this.checkAccess();
    }

    public BrowsePage(IModel<T> model)
    {
        super(model);
        this.checkAccess();
    }

    private void checkAccess()
    {
        GitkiWebSession gitkiWebSession = GitkiWebSession.get();
        if (!gitkiWebSession.isSignedIn()) {
            gitkiWebSession.assertAnonymousBrowsing(BrowsePage.class);
        }
    }

    @Override
    protected IModel<String> createTitleModel()
    {
        return new AbstractPathAbsoluteStringModel(this.getModel());
    }

    @Override
    protected void onInitialize()
    {
        super.onInitialize();

        RepeatingView primaryButtonView = new RepeatingView("primaryButton");
        this.populatePrimaryButtons(primaryButtonView);
        this.add(primaryButtonView);

        this.add(new PathBreadcrumb<>("breadcrumb", this.getModel()));
    }

    protected void populatePrimaryButtons(RepeatingView view)
    {
        /* Hook */
    }

    public GitService getGitService()
    {
        return this.gitService;
    }
}

File: src/main/kotlin/net/dontdrinkandroot/gitki/wicket/page/BrowsePage.kt
package net.dontdrinkandroot.gitki.wicket.page

import net.dontdrinkandroot.gitki.model.GitkiPath
import net.dontdrinkandroot.gitki.service.git.GitService
import net.dontdrinkandroot.gitki.wicket.component.PathBreadcrumb
import net.dontdrinkandroot.wicket.kmodel.KModel
import net.dontdrinkandroot.wicket.kmodel.chain
import net.dontdrinkandroot.wicket.kmodel.kModel
import org.apache.wicket.markup.repeater.RepeatingView
import org.apache.wicket.request.mapper.parameter.PageParameters
import org.apache.wicket.spring.injection.annot.SpringBean

open class BrowsePage<T : GitkiPath> : DecoratorPage<T> {

    @SpringBean
    protected lateinit var gitService: GitService

    constructor(parameters: PageParameters) : super(parameters)

    constructor(model: KModel<T>) : super(model)

    override fun createTitleModel() = kModel.chain { it.absoluteString }

    override fun onInitialize() {
        super.onInitialize()
        val primaryButtonView = RepeatingView("primaryButton")
        populatePrimaryButtons(primaryButtonView)
        this.add(primaryButtonView)
        this.add(PathBreadcrumb("breadcrumb", kModel))
    }

    protected open fun populatePrimaryButtons(view: RepeatingView) {
        /* Hook */
    }
}